File: go/httpmw/options_test.go
package httpmw

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"

	"github.com/sns45/anyonce/go/store/memory"
)

func TestOptions(t *testing.T) {
	t.Run("REQ-HTTP-1: the default methods are POST and PATCH and Methods is matched uppercased", func(t *testing.T) {
		r := Options{}.resolve()
		if !r.methods[http.MethodPost] || !r.methods[http.MethodPatch] || r.methods[http.MethodGet] {
			t.Fatalf("methods %v", r.methods)
		}
		if got := (Options{Methods: []string{"put"}}).resolve().methods; !got["PUT"] || got["POST"] {
			t.Fatalf("methods %v", got)
		}
	})
	t.Run("REQ-HTTP-2: the header name defaults to Idempotency-Key", func(t *testing.T) {
		if r := (Options{}).resolve(); r.HeaderName != "Idempotency-Key" {
			t.Fatal(r.HeaderName)
		}
	})
	t.Run("REQ-HTTP-4: key syntax defaults to lenient", func(t *testing.T) {
		if r := (Options{}).resolve(); r.KeySyntax != "lenient" {
			t.Fatal(r.KeySyntax)
		}
	})
	t.Run("REQ-HTTP-6: fingerprint defaults to body and MaxRequestBytes to 1 MiB", func(t *testing.T) {
		r := (Options{}).resolve()
		if r.Fingerprint != FingerprintBody || r.MaxRequestBytes != 1<<20 {
			t.Fatalf("%+v", r.Options)
		}
	})
	t.Run("REQ-HTTP-8: the store header allowlist defaults to five canonical names", func(t *testing.T) {
		r := (Options{}).resolve()
		for _, name := range []string{"Content-Type", "Content-Language", "Location", "Etag", "Link"} {
			if !r.storeHeaders[name] {
				t.Fatalf("missing %s in %v", name, r.storeHeaders)
			}
		}
		if got := (Options{StoreHeaders: []string{"x-trace"}}).resolve().storeHeaders; !got["X-Trace"] || got["Content-Type"] {
			t.Fatalf("%v", got)
		}
	})
	t.Run("REQ-HTTP-13: the problem base URI and docs URL have D11 defaults", func(t *testing.T) {
		r := (Options{}).resolve()
		if r.ProblemBaseURI != "https://in8.sh/anyonce/problems/" || r.DocsURL != "https://in8.sh/anyonce/problems/missing-key" {
			t.Fatalf("%+v", r.Options)
		}
		if got := (Options{ProblemBaseURI: "https://p.test/"}).resolve().DocsURL; got != "https://p.test/missing-key" {
			t.Fatal(got)
		}
	})
	// This file is the internal test package, so the exported surface is named without the httpmw qualifier.
	t.Run("REQ-WH-2: ProblemTitles reaches the problem the middleware renders", func(t *testing.T) {
		mw := New(memory.New(), Options{
			Required:      true,
			ProblemTitles: map[Code]string{CodeMissingKey: "The webhook-id header is required for this request"},
		})
		rec := httptest.NewRecorder()
		req := httptest.NewRequest(http.MethodPost, "/hook", strings.NewReader("x"))
		mw.Handler(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) { w.WriteHeader(200) })).ServeHTTP(rec, req)
		if rec.Code != 400 {
			t.Fatalf("status = %d, want 400", rec.Code)
		}
		var p Problem
		if err := json.Unmarshal(rec.Body.Bytes(), &p); err != nil {
			t.Fatal(err)
		}
		if p.Title != "The webhook-id header is required for this request" || p.Code != CodeMissingKey {
			t.Fatalf("problem = %+v", p)
		}
	})
}

File: packages/core/test/http/options.test.ts
import { describe, expect, test } from 'bun:test';
import {
  DEFAULT_HEADER_NAME,
  DEFAULT_MAX_REQUEST_BYTES,
  DEFAULT_METHODS,
  DEFAULT_STORED_HEADERS,
  resolveHttpOptions,
} from '../../src/http/options';
import { MemoryStore } from '../../src/memory';

describe('resolveHttpOptions', () => {
  const store = new MemoryStore();

  test('REQ-HTTP-1: the default methods are POST and PATCH, matched case-insensitively', () => {
    expect(DEFAULT_METHODS).toEqual(['POST', 'PATCH']);
    const resolved = resolveHttpOptions({ store, methods: ['put', 'Delete'] });
    expect([...resolved.methods]).toEqual(['PUT', 'DELETE']);
    expect([...resolveHttpOptions({ store }).methods]).toEqual(['POST', 'PATCH']);
  });

  test('REQ-HTTP-2: the header name defaults to Idempotency-Key', () => {
    expect(DEFAULT_HEADER_NAME).toBe('Idempotency-Key');
    expect(resolveHttpOptions({ store }).headerName).toBe('Idempotency-Key');
    expect(resolveHttpOptions({ store, headerName: 'X-Request-Key' }).headerName).toBe(
      'X-Request-Key',
    );
  });

  test('REQ-HTTP-3: required defaults to false', () => {
    expect(resolveHttpOptions({ store }).required).toBe(false);
  });

  test('REQ-HTTP-4: key syntax defaults to lenient', () => {
    expect(resolveHttpOptions({ store }).keySyntax).toBe('lenient');
    expect(resolveHttpOptions({ store, keySyntax: 'strict' }).keySyntax).toBe('strict');
  });

  test('REQ-HTTP-5: requirePrincipal without a principal function throws at construction', () => {
    expect(() => resolveHttpOptions({ store, requirePrincipal: true })).toThrow(/requirePrincipal/);
    expect(
      resolveHttpOptions({ store, requirePrincipal: true, principal: () => 'p' }).requirePrincipal,
    ).toBe(true);
  });

  test('REQ-HTTP-6: fingerprint defaults to body and maxRequestBytes to 1 MiB', () => {
    expect(DEFAULT_MAX_REQUEST_BYTES).toBe(1_048_576);
    const resolved = resolveHttpOptions({ store });
    expect(resolved.fingerprint).toBe('body');
    expect(resolved.maxRequestBytes).toBe(1_048_576);
  });

  test('REQ-HTTP-8: the stored header allowlist defaults to five headers, lowercased', () => {
    expect(DEFAULT_STORED_HEADERS).toEqual([
      'Content-Type',
      'Content-Language',
      'Location',
      'ETag',
      'Link',
    ]);
    expect([...resolveHttpOptions({ store }).storeHeaders]).toEqual([
      'content-type',
      'content-language',
      'location',
      'etag',
      'link',
    ]);
    expect([...resolveHttpOptions({ store, storeHeaders: ['X-Trace'] }).storeHeaders]).toEqual([
      'x-trace',
    ]);
  });

  test('REQ-HTTP-13: the problem base URI and docs URL have D11 defaults', () => {
    const resolved = resolveHttpOptions({ store });
    expect(resolved.problemBaseUri).toBe('https://in8.sh/anyonce/problems/');
    expect(resolved.docsUrl).toBe('https://in8.sh/anyonce/problems/missing-key');
    expect(resolveHttpOptions({ store, problemBaseUri: 'https://p.test/' }).docsUrl).toBe(
      'https://p.test/missing-key',
    );
  });

  test('REQ-HTTP-12: engine policy fields flow into the policy with 3.3 defaults', () => {
    const hooks = {};
    const resolved = resolveHttpOptions({ store, ttlMs: 2000, onStoreError: 'fail-open', hooks });
    expect(resolved.policy.ttlMs).toBe(2000);
    expect(resolved.policy.leaseMs).toBe(30_000);
    expect(resolved.policy.maxResultBytes).toBe(1_048_576);
    expect(resolved.policy.onStoreError).toBe('fail-open');
    expect(resolved.policy.hooks).toBe(hooks);
    expect(resolveHttpOptions({ store }).policy.onStoreError).toBe('fail-closed');
  });
});
